NXDOC-2102: Update NEV deployment documentation (#1533)
nuxeo/doc.nuxeo.com-content: src/nxdoc/nuxeo-add-ons/nuxeo-enhanced-viewer.md

You can see the annotations linked to each version on the same screen, and even annotate one file from this view.

## Installation / Configuration

We assume you have a working Nuxeo on which you have installed the [Nuxeo Enhanced Viewer Connector](https://connect.nuxeo.com/nuxeo/site/marketplace/package/nuxeo-arender) package.

> {{#> callout type='warning' heading='Private artefacts'}}

Reviewer: ```suggestion
{{#> callout type='warning' heading='Private packages'}}
```

> You should contact your Nuxeo Administrator or your Nuxeo sales representative to get access to these artefacts.

Reviewer: ```suggestion
You should contact your Nuxeo Administrator or Nuxeo sales representative to get access to these packages.
```

> We assume you have a working Nuxeo on which you have installed the [Nuxeo Enhanced Viewer Connector](https://connect.nuxeo.com/nuxeo/site/marketplace/package/nuxeo-arender) add-on.

Reviewer 2: ```suggestion
We assume you have a working Nuxeo on which you have installed the [Nuxeo Enhanced Viewer Connector](https://connect.nuxeo.com/nuxeo/site/marketplace/package/nuxeo-arender) package.
```
?

{{/callout}}

### Nuxeo Configuration

The package requires you to define several properties in your `nuxeo.conf`.

#### Before 10.3

> The Nuxeo Enhanced Viewer relies on a [JWT](https://jwt.io/) token to request an OAuth 2 token for authentication (See [OAuth 2]({{page page='using-oauth2'}}) documentation page for more information). Thus, you need to define a JWT secret in your `nuxeo.conf`, which will be used to generate a JWT token.

Reviewer 2: ```suggestion
The Nuxeo Enhanced Viewer relies on a [JWT](https://jwt.io/) token to request an OAuth 2.0 token for authentication (See [OAuth 2.0]({{page page='using-oauth2'}}) documentation page for more information). Thus, you need to define a JWT secret in your `nuxeo.conf`, which will be used to generate a JWT token.
```

> The add-on requires you to define the several properties in your `nuxeo.conf`.

Reviewer 2: ```suggestion
The add-on requires you to define several properties in your `nuxeo.conf`.
```

> You also need to configure an OAuth2 shared secret, between the two applications, with the property `nuxeo.arender.secret` in your `nuxeo.conf`.

Reviewer 2: ```suggestion
You also need to configure an OAuth 2.0 shared secret, between the two applications, with the property `nuxeo.arender.secret` in your `nuxeo.conf`.
```

> In the end, you should have the following in your `nuxeo.conf`:

Reviewer: ```suggestion
In the end, you should have the following properties in your `nuxeo.conf`:
```

```properties
arender.server.previewer.host=https://arender-previewer-url
nuxeo.jwt.secret=JWT_SECRET
nuxeo.arender.secret=OAUTH2_SECRET
```

#### Since 10.3

> The Nuxeo Enhanced Viewer relies on a complete OAuth2 challenge for authentication (See [OAuth 2]({{page page='using-oauth2'}}) documentation page for more information). Thus you need to define an OAuth2 client, with shared secret and redirect URI, under the `nuxeo.arender.oauth2.client.` prefix.

Reviewer 2: ```suggestion
The Nuxeo Enhanced Viewer relies on a complete OAuth 2.0 challenge for authentication (See [OAuth 2.0]({{page page='using-oauth2'}}) documentation page for more information). Thus you need to define an OAuth 2.0 client, with shared secret and redirect URI, under the `nuxeo.arender.oauth2.client.` prefix.
```

> The Nuxeo Enhanced Viewer relies on a complete OAuth 2.0 challenge for authentication (See [OAuth 2.0]({{page page='using-oauth2'}}) documentation page for more information). Thus you need to define an OAuth 2.0 client, with shared secret and redirect URI, under the `nuxeo.arender.oauth2.client.` prefix.

Reviewer: ```suggestion
The Nuxeo Enhanced Viewer relies on a complete OAuth 2.0 challenge for authentication (See [OAuth 2.0]({{page page='using-oauth2'}}) documentation page for more information). Thus, you need to define an OAuth 2.0 client, with a shared secret and a redirect URI, under the `nuxeo.arender.oauth2.client.` prefix.
```

> In the end, you should have the following in your `nuxeo.conf`:

Reviewer 2: ```suggestion
In the end, you should have the following properties in your `nuxeo.conf`:
```
?

> #### With and after 10.3

Reviewer 2: ```suggestion
#### Since 10.3
```
?

```properties
arender.server.previewer.host=https://arender-previewer-url
nuxeo.arender.oauth2.client.create=true
nuxeo.arender.oauth2.client.id=arender
nuxeo.arender.oauth2.client.secret=OAUTH2_SECRET
nuxeo.arender.oauth2.client.redirectURI=/login/oauth2/code/nuxeo
```

### ARender Configuration

> Nuxeo Enhanced Viewer involves to install the ARender services. You can install the ARender services using Kubernetes and Helm 3 by following [ARender Documentation](https://arender.io/v4/documentation/install/kubernetes/).

Reviewer: ```suggestion
Nuxeo Enhanced Viewer involves installing the ARender services. You can install the ARender services using Kubernetes and Helm 3 by following the [ARender Documentation](https://arender.io/v4/documentation/install/kubernetes/).
```

> {{#> callout type='warning' heading='Private artefacts'}}

Reviewer: ```suggestion
{{#> callout type='warning' heading='Private services'}}
```

> You should contact your Nuxeo Administrator or your Nuxeo sales representative to get access to these artefacts.

Reviewer: ```suggestion
You should contact your Nuxeo Administrator or Nuxeo sales representative to get access to these services.
```

{{/callout}}

#### Common

> After getting the ARender Helm Chart, you need to customize some part in order to deploy the Nuxeo connector and add configuration. You can find below an example of _values.yaml_:

Reviewer: ```suggestion
After getting the ARender Helm Chart, you need to customize some parts in order to deploy the Nuxeo connector and add configuration. You can find below an example of _values.yaml_:
```

```yaml
web-ui:
  image:
    repository: docker-arender.packages.nuxeo.com/nuxeo/arender-previewer
    tag: [NEV_VERSION]
  environment:
    # required values

>     ARENDERSRV_NUXEO_SERVER_URL: https://nuxeo-host

Reviewer 2: Worth mentioning the HTTP port in the server URL maybe?

    ARENDERSRV_NUXEO_SERVER_ARENDER_SECRET: OAUTH2_SECRET # Same than the one in nuxeo.conf
    # optional values with their defaults
    ARENDERSRV_NUXEO_CLIENT_TIMEOUT: 30

>     ARENDERSRV_NUXEO_SERVER_CONTEXT_PATH: ''

Reviewer 2: Is it really needed? the context path is already in `ARENDERSRV_NUXEO_SERVER_URL` (?)

Author: It's an optional value to fill. Depending on your setup you may need this. It's our case in the CI because the URL is automatically injected by the exposecontroller without the context path.
Do you think it's better to not mention it?

Author: `ARENDERSRV_NUXEO_SERVER_CONTEXT_PATH` has now `/nuxeo` as default, documentation has been updated to reflect this change.

rendition:
  broker:
    image:
      repository: docker-arender.packages.nuxeo.com/arender-document-service-broker
      tag: [ARENDER_VERSION]
  converter:
    image:
      repository: docker-arender.packages.nuxeo.com/arender-document-converter
      tag: [ARENDER_VERSION]
    environment:
      TKC_TOOLS_IMAGEMAGICK_OPTIONS: "-quality,100,-density,72x72,-units,PixelsPerInch,-auto-orient,+repage"
      TKC_IMAGE_CONVERSION_TIMEOUT: "120"
  dfs:
    image:
      repository: docker-arender.packages.nuxeo.com/arender-document-file-storage
      tag: [ARENDER_VERSION]
  renderer:
    image:
      repository: docker-arender.packages.nuxeo.com/arender-document-renderer
      tag: [ARENDER_VERSION]
  handler:
    image:
      repository: docker-arender.packages.nuxeo.com/arender-document-text-handler
      tag: [ARENDER_VERSION]
```

{{#> callout type='warning' heading='Renaming in 10.3'}}

Before 10.3:

- `ARENDERSRV_NUXEO_SERVER_URL` was named `ARENDERSRV_NUXEO_URL`

> - `ARENDERSRV_NUXEO_SERVER_ARENDER_SECRET` was named `ARENDERSRV_NUXEO_ARENDER_SECRET`

Reviewer 2: Worth keeping backward compatibility?

Author: There's backward compatibility in 10.3.x.

As the 10.2 will be end of life soon, I wanted to make the configuration for 10.3 the first citizen.

{{/callout}}

You can find a release matrix in the repository's [wiki](https://github.com/nuxeo/nuxeo-arender-connector/wiki/Release-Matrix).

#### Ingress

> The Helm Chart doesn't deploy an _Ingress_ by default, you can enable it and customize it regarding your setup with the following:

Reviewer: ```suggestion
The Helm Chart doesn't deploy an _Ingress_ by default, you can enable it and customize it regarding your setup with the following values:
```

```yaml
web-ui:
  ingress:
    enabled: true
    annotations: {}
      # kubernetes.io/ingress.class: nginx
      # kubernetes.io/tls-acme: "true"
    hosts:
      - host: chart-example.local
        paths: []
   tls: []
    #  - secretName: chart-example-tls
    #    hosts:
    #      - chart-example.local
```

#### High-Availability

> You can configure the ARender stack to make it highly available, for that, a couple of settings needs to be added and a database backend is required in order to store OAuth2 tokens and HTTP sessions.

Reviewer 2: ```suggestion
You can configure the ARender stack to make it highly available, for that, a couple of settings needs to be added and a database backend is required in order to store OAuth 2 .0tokens and HTTP sessions.
```

> Currently, the only available backend is MongoDB, the HTTP sessions and OAuth2 access and refresh tokens are stored in the `sessions` and `oauth2Authorized` collections.

Reviewer 2: ```suggestion
Currently, the only available backend is MongoDB, the HTTP sessions and OAuth 2.0 access and refresh tokens are stored in the `sessions` and `oauth2Authorized` collections.
```

--}}
![arender-customized.png](nx_asset://a26f7eaf-e268-443f-be33-2c88542e2a13 ?w=600,border=true)

> The UI customization is done through two files to put in `/docker-entrypoint-init.d` directory inside the Previewer container.

Reviewer: ```suggestion
The UI customization is done through two files to be put in the `/docker-entrypoint-init.d` directory inside the Previewer container.
```

- `arender.css` to customize the style
- `arender.properties` to customize the behavior

Please follow the [ARender configuration guide](https://arender.io/doc/current4/documentation/hmi/index-hmi.html) for more information about available properties and style.

{{#> callout type='info' heading='Properties as environment variables'}}
UI properties can also be customized through environment variables. You need to capitalize all letters in the key, and to replace `.` by `_`, and prefix it with `ARENDER_`.

> For instance,  `topPanel.widgets.beanNames` will become `ARENDER_TOPPANEL_WIDGETS_BEANNAMES`.

Reviewer 2: ```suggestion
For instance, `topPanel.widgets.beanNames` will become `ARENDER_TOPPANEL_WIDGETS_BEANNAMES`.
```
